### DISCUSSION TITLE
🏗️ REST APIの設計について

### DISCUSSION THREAD

**Discussion by kagomen** [lgtm-factory/lgtm-factory#88] [Scrap]
ざっと、調べる！

**Comment by kagomen** [comment]
とりあえずURLの設計について調べる

https://www.strategit.jp/column/041304/

> URLの設計
リソースを洗い出したら次はURLの設計に入ります。 URLはさまざまな要素に関わるため慎重に設計しなければなりません。
>
> URLが設計できたら今後WebAPIの拡張性を高められるように、仕様が理解しやすいものにしておくことが重要です。
>
> URL設計で気をつけておくべき点は下記の通りです。
>
> ひと目でAPIと分かるURLにする
URLにAPIのバージョンを含める
URLには動詞を含めず複数形の名詞だけで構成する
アプリケーションにファイルの拡張子は含めない
リソースの関連性がひと目で分かること
誰が見てもAPIと分かることは非常に大切で、例としてはディレクトリに分ける方法とサブドメインに分ける方法があります。
>
> URLにAPIバージョンを含めればREST APIを使用する開発者がAPIのバージョンを選択しやすく、切り分けもしやすくなります。
>
> URLはリソースそのものを表し、操作についてはHTTPメソッドで表現することが必要です。 URLの末尾には言語に関連した拡張子を付けません。
>
> それは将来的にAPIの仕様が変更された際に容易に変更作業がしやすくなるからです。 また、URLは実装が手間になるため長すぎないようにしてください。
>
> URLは見ただけでリソースがどこのものか分かるようにしておくといちいち調べる手間が省けます。

**Reply by kagomen** [reply]
https://learn.microsoft.com/ja-jp/azure/architecture/best-practices/api-design#organize-the-api-design-around-resources

> 可能であれば、リソース URI は動詞 (リソースに対する操作) ではなく、名詞 (リソース) に基づくようにします。
>
> リソースは、単一の物理データ項目に基づく必要はありません。 たとえば、注文リソースをリレーショナル データベースの複数のテーブルとして内部的に実装しながら、クライアントには単一のエンティティとして表示することができます。 データベースの内部構造を単にミラー化する API は作成しないようにします。 REST の目的は、エンティティと、アプリケーションがそれらのエンティティに対して実行できる操作をモデル化することです。 クライアントを内部実装に公開しないでください。
>
> 多くの場合、エンティティはコレクション (orders、customers) にグループ化されます。 コレクションはコレクションの項目とは別のリソースであり、独自の URI が必要です。 
>
> URI には一貫性のある名前付け規則を使用します。 一般に、コレクションを参照する URI には複数形の名詞を使用すると便利です。 コレクションと項目の URI は、階層構造に整理することをお勧めします。 たとえば、 /customers は customers コレクションのパスであり、 /customers/5 は ID が 5 の顧客のパスです。 この方法は、Web API を直感的に保つのに役立ちます。 また、多くの Web API フレームワークでは、パラメーター化された URI パスに基づいて要求をルーティングできるので、パス /customers/{id} のルートを定義できます。

**Reply by kagomen** [reply]
なるほど！そうすると、api/v1/images?theme={theme}が妥当、かな？

**Reply by kagomen** [reply]
画像データの置き場所がlgtm-dataなら、これと統一したいところ

- api/v1/lgtms
- api/v1/lgtm-images

などなど

api/v1/lgtm-images　が良い気がする！

**Reply by kagomen** [reply]
dataの複数形はdataなんだ

**Comment by kagomen** [comment]
プログラミングでよく使う英単語のまとめ【随時更新】
https://qiita.com/Ted-HM/items/7dde25dcffae4cdc7923

**Comment by kagomen** [comment]
以下と、

```ts
// api/v1/lgtm-images/route.tsx

import { NextRequest } from "next/server";

export const runtime = "edge";

export async function GET(request: NextRequest) {
  const searchParams = request.nextUrl.searchParams;
  const theme = searchParams.get("theme");

  const { lgtmImage } = await import(`../../../../../lgtm-data/${theme}`);

  return lgtmImage();
}


// lgtm-data/test.tsx

export const lgtmImage = () => {
  return new ImageResponse(
    (
      <div
        style={{
          display: "flex",
          height: "100%",
          width: "100%",
          alignItems: "center",
          justifyContent: "center",
          background: "white",
          fontSize: "64px",
        }}
      >
        LGTM🐈‍⬛
      </div>
    ),
    {
      width: 1200,
      height: 630,
    },
  );
};
```

**Reply by kagomen** [reply]
以下

```ts
import { NextRequest } from "next/server";
import { ImageResponse } from "next/og";

export const runtime = "edge";

export async function GET(request: NextRequest) {
  const searchParams = request.nextUrl.searchParams;
  const theme = searchParams.get("theme");

  const { fetchLgtmData } = await import(`../../../../../lgtm-data/${theme}`);
  const { jsx, options } = fetchLgtmData()

  return new ImageResponse(jsx, options)
}

export const fetchLgtmData = () => {
  const options = {
    width: 1200,
    height: 630,
  };
  const jsx = (
    <div
      style={{
        display: "flex",
        height: "100%",
        width: "100%",
        alignItems: "center",
        justifyContent: "center",
        background: "white",
        fontSize: "64px",
      }}
    >
      LGTM🐈‍⬛
    </div>
  );
  return {
    jsx,
    options,
  };
};
```

**Reply by kagomen** [reply]
@kazzyfrog 
前者は処理の流れが追いやすく、ユーザーがlgtm-dataを新しく作るにも、ImageResponseのドキュメントと形式が一致しているため理解しやすいかな？と思うのですが、どう思いますか？🤔

前者より後者の方がパフォーマンスに優れる、といったようなことなどがありますか・・・？👀

**Reply by kazzyfrog** [reply]
APIエンドポイント自体、結果をキャッシュできると思うので、
パフォーマンス面は、同じかなと思います！👍

個人的には、
ユーザーは、next/ImageResponse の機能まで知らなくてもいいのかな？と思った感じです！

少しでもなんか、簡単に参加できるようにしたほうがいいのかなと！

なので、単純に、JSXと、そのオプションをいじれば作成できるように、
ImageResponse の機能は隠してもいいのかも、、？

同じ理由で、そのディレクトリだけ、tsxではなく、jsxにしたほうがいいかもと思ってます！

例えば、こんな感じのテンプレファイルを用意しとく感じですかね！
```jsx
export const fetchLgtmData = () => {

  const options = {
  // ここに、作成するLGTM画像のオプションを記述します
    width: 1200,
    height: 630,
  };

  const jsx = (
  // ここに、作成するLGTM画像をコーディングします。
  // nom run dev で、プレビューしながら、下記の初期のコードは削除してください。
  // Tailwind CSSを使う場合は、〜
    <div
      style={{
        display: "flex",
        height: "100%",
        width: "100%",
        alignItems: "center",
        justifyContent: "center",
        background: "white",
        fontSize: "64px",
      }}
    >
      LGTM🐈‍⬛
    </div>
  );

  return {
    jsx,
    options,
  };
};
```

**Reply by kazzyfrog** [reply]
あと、単純に、ImageResponseのロジックは１箇所で管理したい気持ちがありますかね！

[バージョン履歴](https://nextjs.org/docs/app/api-reference/functions/image-response#version-history)見るとわかる通り、
例えば、バージョンアップで、インポート元が変わったりしたら、前ファイル修正必要になってしまうかも、なので👍

後者の場合、ユーザーが作成するファイルの内容は、
何もインポートしない、staticなデータになると思うので、そのほうが管理しやすいかなと！

要は、
lgtm-dataディレクトリは、あくまでデータファイルで、ロジックは持たせないほうが、管理しやすいかなと！
UIコンポーネントのDB的な役割をさせる、ディレクトリですかね！

**Reply by kagomen** [reply]
> [バージョン履歴](https://nextjs.org/docs/app/api-reference/functions/image-response#version-history)見るとわかる通り、
例えば、バージョンアップで、インポート元が変わったりしたら、前ファイル修正必要になってしまうかも

> 後者の場合、ユーザーが作成するファイルの内容は、何もインポートしない、staticなデータになると思うので、そのほうが管理しやすい

なるほどーー！！
すごい！そういう理由があったんですね！！🌟
ご説明ありがとうございます！すべて了解しました！🫡🌈
